contracts/stake-cw721/src/contract.rs

```diff
@@ -0,0 +1,1844 @@
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    let owner = match msg.owner {
+        Some(owner) => Some(deps.api.addr_validate(owner.as_str())?),
+        None => None,
+    };
+
+    let manager = match msg.manager {
+        Some(manager) => Some(deps.api.addr_validate(manager.as_str())?),
+        None => None,
+    };
+
+    let reward_token = match msg.reward_token_address {
+        Some(reward_token) => Some(deps.api.addr_validate(&*reward_token)?),
+        None => None,
+    };
+
+    let config = Config {
+        owner,
+        manager,
+        nft_address: deps.api.addr_validate(&*msg.nft_address)?,
+        reward_token_address: reward_token,
+        unstaking_duration: msg.unstaking_duration,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<Empty>, ContractError> {
+    match msg {
+        ExecuteMsg::Receive(msg) => execute_fund(deps, env, info, msg),
+        ExecuteMsg::ReceiveNft(msg) => execute_stake(deps, env, info, msg),
+        ExecuteMsg::Unstake {
+            token_id,
+            reward_wallet_address,
+        } => execute_unstake(deps, env, info, token_id, reward_wallet_address),
+        ExecuteMsg::ClaimNfts {} => execute_claim_nfts(deps, env, info),
+        ExecuteMsg::ClaimRewards {} => execute_claim_rewards(deps, env, info),
+        ExecuteMsg::UpdateConfig {
+            owner,
+            manager,
+            duration,
+        } => execute_update_config(info, deps, owner, manager, duration),
+        ExecuteMsg::AddHook { addr } => execute_add_hook(deps, env, info, addr),
+        ExecuteMsg::RemoveHook { addr } => execute_remove_hook(deps, env, info, addr),
+    }
+}
+
+pub fn execute_fund(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    wrapper: Cw20ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if let Some(reward_token_address) = config.reward_token_address {
+        let reward_token_address = deps.api.addr_validate(reward_token_address.as_ref())?;
+        if info.sender != reward_token_address {
+            return Err(ContractError::InvalidToken {
+                received: info.sender,
+                expected: reward_token_address,
+            });
+        }
+
+        let sender = deps.api.addr_validate(&wrapper.sender)?;
+        let balance = REWARD_BALANCE.load(deps.storage).unwrap_or_default();
+        REWARD_BALANCE.save(
+            deps.storage,
+            &balance
+                .checked_add(wrapper.amount)
+                .map_err(StdError::overflow)?,
+        )?;
```

Comment:
Think the `update` method does what you want here. Something like:
```suggestion
        REWARD_BALANCE.update(deps.storage, |old| old.unwrap_or_default().checked_add(wrapper.amount).map_err(StdError::overflow))?;
```

Reply:
Cool.